// tmi.js/lib/logger.js
var _ = require('./utils')

var currentLevel = 'info'
var levels = { 'trace': 0, 'debug': 1, 'info': 2, 'warn': 3, 'error': 4, 'fatal': 5 }

// Logger implementation..
function log (level) {
  // Return a console message depending on the logging level..
  return function (message) {
    // if (levels[level] >= levels[currentLevel]) {
    //     console.log(`[${_.formatDate(new Date())}] ${level}: ${message}`)
    // }

    // console.log(message)

    // if (message.indexOf('spddl_bot') !== -1) message = message.replace(/spddl_bot/g, '\x1B[32mspddl_bot\x1B[39m')
    // if (message.indexOf('spddl') !== -1) message = message.replace(/spddl/g, '\x1B[32mspddl\x1B[39m')

    if (message.indexOf('<spddl_bot>') !== -1) message = message.replace(/<spddl_bot>/g, '<\x1B[32mspddl_bot\x1B[39m>')
    if (message.indexOf('<spddl>') !== -1) message = message.replace(/<spddl>/g, '<\x1B[32mspddl\x1B[39m>')

    console.log((global.bot.limit >= 1 ? '\x1b[35m[' + global.bot.limit + ']\x1b[39m ' : '') + message)
  }
}

module.exports = {
  // Change the current logging level..
  setLevel: function (level) {
    currentLevel = level
  },
  trace: log('trace'),
  debug: log('debug'),
  info: log('info'),
  warn: log('warn'),
  error: log('error'),
  fatal: log('fatal')
}
